internal/core/src/exec/Driver.cpp

```diff
@@ -0,0 +1,350 @@
+    auto self = shared_from_this();
+
+    RowVectorPtr result;
+    auto stop = RunInternal(self, blocking_state, result);
+
+    Assert(stop == StopReason::kBlock || stop == StopReason::kAtEnd ||
+           stop == StopReason::kAlreadyTerminated);
+    return result;
+}
+
+#define CALL_OPERATOR(call_func, operator, method_name)                        \
+    try {                                                                      \
+        call_func;                                                             \
+    } catch (std::exception & e) {                                             \
+        throw ExecOperatorException(                                           \
+            fmt::format("Operator::{} failed for [Operator:{}, plan node id: " \
+                        "{}] : {}",                                            \
+                        method_name,                                           \
+                        operator->get_operator_type(),                         \
+                        operator->get_plannode_id(),                           \
+                        e.what()));                                            \
+    }
+
+StopReason
+Driver::RunInternal(std::shared_ptr<Driver>& self,
+                    std::shared_ptr<BlockingState>& blocking_state,
+                    RowVectorPtr& result) {
+    try {
+        int num_operators = operators_.size();
+        ContinueFuture future;
+
+        for (;;) {
+            for (size_t i = num_operators - 1; i >= 0; --i) {
+                auto op = operators_[i].get();
+
+                current_operator_index_ = i;
+                CALL_OPERATOR(
+                    blocking_reason_ = op->IsBlocked(&future), op, "IsBlocked");
+                if (blocking_reason_ != BlockingReason::kNotBlocked) {
+                    blocking_state = std::make_shared<BlockingState>(
+                        self, std::move(future), op, blocking_reason_);
+                    return StopReason::kBlock;
+                }
+                Operator* next_op = nullptr;
+
+                if (i < operators_.size() - 1) {
+                    next_op = operators_[i + 1].get();
+                    CALL_OPERATOR(
+                        blocking_reason_ = next_op->IsBlocked(&future),
+                        next_op,
+                        "IsBlocked");
+                    if (blocking_reason_ != BlockingReason::kNotBlocked) {
+                        blocking_state = std::make_shared<BlockingState>(
+                            self, std::move(future), next_op, blocking_reason_);
+                        return StopReason::kBlock;
+                    }
+
+                    bool needs_input;
+                    CALL_OPERATOR(needs_input = next_op->NeedInput(),
+                                  next_op,
+                                  "NeedInput");
+                    if (needs_input) {
+                        RowVectorPtr result;
```

Comment:
rename variable, this local variable shadows one of the method args, which could be confusing.